tests/engine/processes/workchains/test_restart.py

```diff
@@ -8,44 +8,94 @@
 
-        assert SomeWorkChain.is_process_handler('handler_a')
-        assert not SomeWorkChain.is_process_handler('not_a_handler')
-        assert not SomeWorkChain.is_process_handler('unexisting_method')
 
-    @staticmethod
-    def test_get_process_handler():
-        """Test the `BaseRestartWorkChain.get_process_handlers` class method."""
+def test_is_process_handler():
+    """Test the `BaseRestartWorkChain.is_process_handler` class method."""
+    assert SomeWorkChain.is_process_handler('handler_a')
+    assert not SomeWorkChain.is_process_handler('not_a_handler')
+    assert not SomeWorkChain.is_process_handler('unexisting_method')
 
-        class SomeWorkChain(BaseRestartWorkChain):
-            """Dummy class."""
 
-            @process_handler
-            def handler_a(self, node):
-                pass
+def test_get_process_handler():
+    """Test the `BaseRestartWorkChain.get_process_handlers` class method."""
+    assert [handler.__name__ for handler in SomeWorkChain.get_process_handlers()] == ['handler_a']
 
-            def not_a_handler(self, node):
-                pass
 
-        assert [handler.__name__ for handler in SomeWorkChain.get_process_handlers()] == ['handler_a']
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_excepted_process(generate_work_chain, generate_calculation_node):
+    """Test that the workchain aborts if the sub process was excepted."""
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.children = [generate_calculation_node(ProcessState.EXCEPTED)]
+    assert process.inspect_process() == BaseRestartWorkChain.exit_codes.ERROR_SUB_PROCESS_EXCEPTED  # pylint: disable=no-member
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_killed_process(generate_work_chain, generate_calculation_node):
+    """Test that the workchain aborts if the sub process was killed."""
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.children = [generate_calculation_node(ProcessState.KILLED)]
+    assert process.inspect_process() == BaseRestartWorkChain.exit_codes.ERROR_SUB_PROCESS_KILLED  # pylint: disable=no-member
+
+
+@pytest.mark.usefixtures('clear_database_before_test')
+def test_unhandled_failure(generate_work_chain, generate_calculation_node):
+    """Test the unhandled failure mechanism.
+
+    The workchain should be aborted if there are two consecutive failed sub processes that went unhandled. We simulate
+    it by setting `ctx.unhandled_failure` to True and append two failed process nodes in `ctx.children`.
+    """
+    process = generate_work_chain(SomeWorkChain, {})
+    process.setup()
+    process.ctx.unhandled_failure = True
+    process.ctx.iteration = 2
+    process.ctx.children = [generate_calculation_node(exit_status=100) for _ in range(2)]
+    assert process.inspect_process() == BaseRestartWorkChain.exit_codes.ERROR_SECOND_CONSECUTIVE_UNHANDLED_FAILURE  # pylint: disable=no-member
+
```

Comment:
Could I ask you how these tests work? It seems like you are getting the `work_chain` process and then manually modifying the context of the workchain from outside of it. As a side note, I don't know why, but I was assuming the context was something that was only modifiable from inside the workchain methods: anyways this is just for testing and that should be the intended way to use it, right?

On the other hand, if you are just modifying things manually I'm not sure how does the workchain error code gets set because I don't see any run or submits, so I only see you setting up the workchain. Is it being sent to run when you `instantiate_process` with the `runner` in `generate_work_chain`? If that is the case, how are you sure that when you modify it the workchain has not already finished? I am not familiar with this way of running stuff...

Reply:
The whole point of these tests is exactly that really "running" them is very difficult. This is because it often requires a lot of pre-requisites. In the case of the `BaseRestartWorkChain`, which I am testing here, it needs to run some `CalcJob`, which in turn needs a `Code` that then requires the corresponding binary to be installed. Since I don't want to test all of this in a big integration test, but want to write smaller unit tests of the individual methods of the `BaseRestartWorkChain` I have to mock certain things. Since the methods I am testing are class methods, I first need to have an instance. This is what the `generate_work_chain` is doing. It just gets me a `WorkChain` instance. There I can then call the methods I want to test. In this case, I want to test the `inspect_process` method. This is normally not the first method that is called, so I have to make sure to correctly "mock" the state of the workchain instance that it would normally have, when run fully automatically by the engine. This includes defining the `ctx` correctly, which is an internal dictionary of the `WorkChain` where it stores temporary data. I populate it such that it should trigger a certain code path in `inspect_process` and check that it behaves correctly.

Reply:
Uhm, I see. I think that my main confussion was that I thought `inspect_process` just got you a result that was set by the engine when running the process, but now I understand this is not so and the `inspect_process` itself looks at the status of the calculations in the context to decide its output. Correct?

Reply:
The `inspect_process` is just a custom `WorkChain` method, that is one of the steps that is called in its outline:
https://github.com/aiidateam/aiida-core/blob/a0e39c69ba7fb15eb36cc400789a172b3670af6d/aiida/engine/processes/workchains/restart.py#L173

So I am just testing the code that is in there.